### ISSUE TITLE
Transformers trainer reporting incorrect loss to comet dashboard

### ISSUE THREAD

**Issue by bhperry** [comet-ml/issue-tracking#523]
## Describe the Bug
Using cometml with huggingface transformers to track fine tuning progress. Every time I train, after a few steps the CometML dashboard will suddenly show loss for the next "10s" step, e.g. here at step 4 it has a loss value for step 10 already.

![loss-from-future-steps](https://github.com/comet-ml/issue-tracking/assets/11621797/92719ae5-5437-4b75-83d3-0e5f04e9c1ae)

And this continues every 10 steps
![loss-v-steps](https://github.com/comet-ml/issue-tracking/assets/11621797/4becb844-2c3f-4d71-b4e7-7872b28876c8)

None of those low loss values are ever reported in the logs, and considering they seem to appear on future steps I can't imagine they are real.

## Expected behavior
Expected not to see incorrect loss values reported for steps that haven't happened yet.

## Where is the issue?
- [x] Comet Python SDK
- [x] Comet UI
- [x] Third Party Integrations (Huggingface, TensorboardX, Pytorch Lighting etc)

## To Reproduce
Configuring the standard huggingface Trainer with `report_to="comet_ml"` in TrainingArguments and running `comet_ml.init(....)` on startup.
1. Init comet-ml `comet_ml.init(api_key=.....)`
2. With standard Huggingface Trainer, set `report_to="comet_ml"` in TrainingArguments
3. Train over a dataset with more than 10 examples
4. Watch comet-ml dashboard for bad loss data

**Comment by Lothiraldan** [comment]
@bhperry Thank you for reporting this issue, we are looking into it. In order to help us reproduce the issue, could you share the following piece of information:
* Output of `pip list`
* Output of `python -V`
* If you could share a link to your experiment or your code, that would be ideal. You can share them privately with me by email at boris@comet.com. If not, could you share your `TrainingArguments` and `Trainer` arguments?

**Comment by bhperry** [comment]
Sure thing, here's the env details and training script

Python version: `Python 3.9.18`

<details><summary>Pip list</summary>

```
Package                       Version
----------------------------- ------------
accelerate                    0.22.0
aiohttp                       3.8.5
aiosignal                     1.3.1
altair                        5.1.1
asttokens                     2.4.0
async-timeout                 4.0.3
attrs                         23.1.0
Automat                       22.10.0
backcall                      0.2.0
backports.functools-lru-cache 1.6.5
beautifulsoup4                4.12.2
bitsandbytes                  0.41.1
blinker                       1.6.2
cachetools                    5.3.1
certifi                       2023.7.22
cffi                          1.15.1
charset-normalizer            3.2.0
click                         8.1.7
cmake                         3.27.4.1
comet-llm                     1.3.1
comet-ml                      3.33.11
comm                          0.1.4
configobj                     5.0.8
constantly                    15.1.0
cryptography                  41.0.3
cssselect                     1.2.0
dataclasses-json              0.5.14
datasets                      2.14.5
debugpy                       1.7.0
decorator                     5.1.1
dill                          0.3.7
dulwich                       0.21.6
einops                        0.6.1
everett                       3.1.0
exceptiongroup                1.1.3
executing                     1.2.0
faiss                         1.7.2
filelock                      3.12.3
flatten-dict                  0.4.2
frozenlist                    1.4.0
fsspec                        2023.6.0
gitdb                         4.0.10
GitPython                     3.1.36
gptq                          0.0.3
greenlet                      2.0.2
huggingface-hub               0.17.1
hyperlink                     21.0.0
idna                          3.4
importlib-metadata            6.8.0
incremental                   22.10.0
intel-openmp                  2021.4.0
ipykernel                     6.13.0
ipython                       8.15.0
ipywidgets                    8.1.0
itemadapter                   0.8.0
itemloaders                   1.1.0
jedi                          0.19.0
Jinja2                        3.1.2
jmespath                      1.0.1
joblib                        1.3.2
jsonschema                    4.19.0
jsonschema-specifications     2023.7.1
jupyter_client                8.3.1
jupyter_core                  5.3.1
jupyterlab-widgets            3.0.8
langchain                     0.0.222
langchainplus-sdk             0.0.20
lit                           16.0.6
lxml                          4.9.3
markdown-it-py                3.0.0
MarkupSafe                    2.1.3
marshmallow                   3.20.1
marshmallow-dataclass         8.5.14
matplotlib-inline             0.1.6
mdurl                         0.1.2
mkl                           2021.4.0
mpmath                        1.3.0
multidict                     6.0.4
multiprocess                  0.70.15
mypy-extensions               1.0.0
nest-asyncio                  1.5.6
networkx                      3.1
nltk                          3.8.1
numexpr                       2.8.6
numpy                         1.24.4
nvidia-cublas-cu11            11.10.3.66
nvidia-cuda-cupti-cu11        11.7.101
nvidia-cuda-nvrtc-cu11        11.7.99
nvidia-cuda-runtime-cu11      11.7.99
nvidia-cudnn-cu11             8.5.0.96
nvidia-cufft-cu11             10.9.0.58
nvidia-curand-cu11            10.2.10.91
nvidia-cusolver-cu11          11.4.0.1
nvidia-cusparse-cu11          11.7.4.91
nvidia-nccl-cu11              2.14.3
nvidia-nvtx-cu11              11.7.91
openapi-schema-pydantic       1.2.4
packaging                     23.1
pandas                        2.1.0
parsel                        1.8.1
parso                         0.8.3
peft                          0.6.0.dev0
pexpect                       4.8.0
pickleshare                   0.7.5
Pillow                        9.5.0
pip                           23.2.1
platformdirs                  3.10.0
prompt-toolkit                3.0.39
Protego                       0.3.0
protobuf                      4.24.3
psutil                        5.9.5
ptyprocess                    0.7.0
pure-eval                     0.2.2
pyarrow                       13.0.0
pyasn1                        0.5.0
pyasn1-modules                0.3.0
pycparser                     2.21
pydantic                      1.10.12
pydeck                        0.8.0
PyDispatcher                  2.0.7
Pygments                      2.16.1
Pympler                       1.0.1
pyOpenSSL                     23.2.0
python-box                    6.1.0
python-dateutil               2.8.2
pytz                          2023.3.post1
pytz-deprecation-shim         0.1.0.post0
PyYAML                        6.0.1
pyzmq                         25.1.1
queuelib                      1.6.2
referencing                   0.30.2
regex                         2023.8.8
requests                      2.31.0
requests-file                 1.5.1
requests-toolbelt             1.0.0
rich                          13.5.2
rpds-py                       0.10.2
safetensors                   0.3.3
saturnfs                      0.0.1
scikit-learn                  1.3.0
scipy                         1.11.2
Scrapy                        2.10.1
semantic-version              2.10.0
sentence-transformers         2.2.2
sentencepiece                 0.1.99
sentry-sdk                    1.31.0
service-identity              23.1.0
setuptools                    68.1.2
simplejson                    3.19.1
six                           1.16.0
smmap                         5.0.0
soupsieve                     2.5
SQLAlchemy                    2.0.20
sseclient                     0.0.27
stack-data                    0.6.2
streamlit                     1.26.0
sympy                         1.12
tbb                           2021.10.0
tenacity                      8.2.3
threadpoolctl                 3.2.0
tldextract                    3.5.0
tokenizers                    0.13.3
toml                          0.10.2
toolz                         0.12.0
torch                         2.0.1
torchvision                   0.15.2
tornado                       6.3.3
tqdm                          4.66.1
traitlets                     5.9.0
transformers                  4.33.1
triton                        2.0.0
Twisted                       22.10.0
types-requests                2.31.0.7
typing_extensions             4.7.1
typing-inspect                0.9.0
tzdata                        2023.3
tzlocal                       4.3.1
urllib3                       2.0.4
validators                    0.22.0
w3lib                         2.1.2
watchdog                      3.0.0
wcwidth                       0.2.6
websocket-client              1.3.3
wheel                         0.41.2
widgetsnbextension            4.0.8
wrapt                         1.15.0
wurlitzer                     3.0.3
xformers                      0.0.21
xxhash                        3.3.0
yarl                          1.9.2
zipp                          3.16.2
zope.interface                6.0
```
</details>

<details><summary>Training Script</summary>

```
import os
import pathlib
from typing import Tuple
import comet_ml

from datasets import load_from_disk
import torch
from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_kbit_training
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, DataCollatorForTokenClassification, PreTrainedTokenizerBase, Trainer, TrainingArguments


NAME = "llama2-7b-chat-lora"
BASE_MODEL = "meta-llama/Llama-2-7b-chat-hf"
MODEL_MAX_LENGTH = 4096

WORLD_SIZE = int(os.getenv("WORLD_SIZE", 1))
LOCAL_RANK = int(os.getenv("LOCAL_RANK", 0))
GLOBAL_RANK = int(os.getenv("RANK", 0))


def has_checkpoint(dir: str) -> bool:
    if list(pathlib.Path(dir).glob("checkpoint-*")):
        return True
    return False


def load_base_model(lora_config: LoraConfig, gradient_checkpointing: bool = True) -> Tuple[PeftModel, PreTrainedTokenizerBase]:
    tokenizer = AutoTokenizer.from_pretrained(
        BASE_MODEL,
        model_max_length=MODEL_MAX_LENGTH,
        padding_side="right",
    )
    tokenizer.pad_token = tokenizer.unk_token

    # Load model to GPU specified by LOCAL_RANK
    device_map = (
        {"": LOCAL_RANK}
    )
    model = AutoModelForCausalLM.from_pretrained(
        BASE_MODEL,
        quantization_config=BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_quant_type="nf4",
            bnb_4bit_compute_dtype=torch.float16,
        ),
        device_map=device_map,
    )
    model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=gradient_checkpointing)

    if torch.cuda.device_count() > 1:
        # keeps Trainer from trying its own DataParallelism when more than 1 gpu is available
        model.is_parallelizable = True
        model.model_parallel = True

    # Load LoRa adapter
    model = get_peft_model(model, lora_config)
    model.config.use_cache = False

    if LOCAL_RANK == 0:
        model.print_trainable_parameters()

    return model, tokenizer


def tune():
    output_dir = os.path.join(os.getenv("LOCAL_MODELS_DIR"), NAME)
    dataset = load_from_disk(os.getenv("DATASET_NAME"))
    train_dataset = dataset["train"]
    eval_dataset = dataset["test"]

    training_args = TrainingArguments(
        output_dir,
        do_train=True,
        do_eval=True,
        optim="adamw_torch",
        lr_scheduler_type="cosine",
        auto_find_batch_size=True,
        gradient_accumulation_steps=8,
        num_train_epochs=2,
        evaluation_strategy="steps",
        eval_steps=10,
        eval_accumulation_steps=16,
        learning_rate=2e-5,
        warmup_ratio=0.001,
        logging_steps=1,
        gradient_checkpointing=True,
        ddp_find_unused_parameters=False,
        report_to="comet_ml",
    )
    lora_config = LoraConfig(
        r=8,
        lora_alpha=16,
        lora_dropout=0.05,
        target_modules=["q_proj", "v_proj"],
        bias="none",
        task_type="CAUSAL_LM",
    )
    model, tokenizer = load_base_model(lora_config, gradient_checkpointing=training_args.gradient_checkpointing)
    data_collator = DataCollatorForTokenClassification(tokenizer)

    trainer = Trainer(
        model,
        args=training_args,
        data_collator=data_collator,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        tokenizer=tokenizer,
    )
    trainer.train()

if __name__ == "__main__":
    comet_ml.init(
        api_key=os.getenv("COMET_API_KEY"),
        project_name=NAME,
        workspace="llm",
    )
    tune()
```
</details>

**Comment by bhperry** [comment]
Training on ms-marco v1.1 QA formatted into the Llama2 style. But I don't think the specifics of the data should matter, I see this behavior on every run.

**Comment by Lothiraldan** [comment]
@bhperry You were right, I think I was able to reproduce the behavior you are seeing with our simpler example https://github.com/comet-ml/comet-examples/blob/master/integrations/model-training/hugging_face/notebooks/Comet_with_Hugging_Face_Trainer.ipynb by using the same training arguments. 

Here is the experiment that exhibit the same charts: https://www.comet.com/lothiraldan/debug-issue-tracking-523/3bd9415f042f425fb0c7409d532a277a?experiment-tab=panels&showOutliers=true&smoothing=0&xAxis=step

We will now working on understanding where the issue come from and come up with a solution

**Comment by bhperry** [comment]
Thanks @Lothiraldan! Appreciate the update.

**Comment by Lothiraldan** [comment]
Hi @bhperry, we have identified the issue. It appears to be a compatibility problem between our PyTorch integration and Accelerate's gradient accumulation feature, which is enabled through the `gradient_accumulation_steps` training argument. This is resulting in incorrect loss values being reported. We are actively working on finding a solution to fix this problem.

In the meantime, you can temporarily disable our PyTorch integration by importing torch before comet_ml. To do this, modify the imports as follows:

```python
import os
import pathlib
from typing import Tuple

import torch

import comet_ml

from datasets import load_from_disk
from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_kbit_training
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, DataCollatorForTokenClassification, PreTrainedTokenizerBase, Trainer, TrainingArguments

```

By making this change, our PyTorch integration will be disabled. You will see the following warning message, which can be ignored: `COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: torch.`. The loss metric will continue to be correctly reported by the Transformers integration.

We apologize for the inconvenience and we will keep you informed about the solution we come up with.

**Comment by dsblank** [comment]
Internal issue number: EXT-1325

**Comment by dsblank** [comment]
Looks like a solution has been included in comet_ml version 3.35.0. Can you give it a try?
